### PR TITLE
Script lookup iframes by frameid

### PR DESCRIPTION
**Pull request by asajeffrey**
Lookup iframes by `FrameId` rather than `PipelineId`. This should make lookup much more reliable, since the frame id doesn't change.

cc @ConnorGBrewster @aneeshusa 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because refactoring

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13646)

### CODE REVIEW

**Comment by asajeffrey** [comment]
The first commit in this PR is actually #13627, we shouldn't merge this one until it lands.

**Comment by asajeffrey** [comment]
Rebased.

File: components/script_traits/script_msg.rs
Comment (nox) [inline components/script_traits/script_msg.rs]: What does it mean that it is "in the frame"? It still lives here so it's a but confusing.

**Review by nox — COMMENTED** [review]
Just a question, otherwise r=me.

File: components/script_traits/script_msg.rs
Comment (asajeffrey) [inline components/script_traits/script_msg.rs]: Good point, that second pipeline id should be a frame id.

**Comment by asajeffrey** [comment]
Working on this, turns out there were bugs, sigh.

**Comment by jdm** [comment]
I haven't done more than skim recent changes in this PR, but I'm concerned about the changes to how the document load event is initiated. The old design was that it was fine to optimistically enqueue a "I think the document is finished loading everything" event, because when being run it would first check if that were actually the case. Some of the changes and comments around them make me worried that this is no longer true.

**Comment by asajeffrey** [comment]
@jdm yes, this is what I've been spending my time debugging. I now understand better what the document loading code is doing, I even wrote some comments so that future me might remember this. At least the wpt test suite caught this.

**Comment by asajeffrey** [comment]
This PR triggered issue #13964, fixed by #13965, which should land before this does.

**Comment by asajeffrey** [comment]
I appear to have stomped the last bug down, @ConnorGBrewster or @nox, care to take a look at this PR again? It's mainly just ab6af830ff794d9f5fbbde7d21ca17258989da3a that's changed.

**Comment by asajeffrey** [comment]
And of course just when I say I got the last bug, another one pops up. The fix is 82b8677535cbe2cd55baf9e833050920673833b7, which reverts back to the old behaviour of only creating frames once the first pipeline has matured.

File: components/constellation/constellation.rs
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: Could this and the code below cause multiple closures of the same pipeline/frame?

File: components/constellation/constellation.rs
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: Change `mozbrowser_iframe_id` to `mozbrowser_frame_id`?

File: components/net/http_loader.rs
Comment (ConnorGBrewster) [inline components/net/http_loader.rs]: Any reason for this being in this PR?

File: components/script/dom/node.rs
Comment (ConnorGBrewster) [inline components/script/dom/node.rs]: Nit:
```
self.inclusive_ancestors().last()
    .and_then(Root::downcast::<Document>)
    .and_then(|doc| doc.browsing_context())
    .is_some()
```

**Review by cbrewster — CHANGES_REQUESTED** [review]
After the initial review, this looks great!
Just a couple nits/questions. It might be worth running this through CI a few times to see if any intermittents are introduced with the logic change for when a document matures.

File: components/constellation/constellation.rs
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Yes. I guess I should add a comment to that effect, that closing a frame or pipeline should be idempotent.

File: components/constellation/constellation.rs
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Done.

File: components/net/http_loader.rs
Comment (asajeffrey) [inline components/net/http_loader.rs]: I was doing some debugging in XHR, and was having a hard time because the url wasn't being logged. Not really core business of this PR, but doing this in its own PR seems a bit much.

File: components/script/dom/node.rs
Comment (asajeffrey) [inline components/script/dom/node.rs]: Doesn't borrow check I'm afraid.

Comment (ConnorGBrewster) [inline components/script/dom/node.rs]: I figured that was probably the reason :(

File: components/net/http_loader.rs
Comment (ConnorGBrewster) [inline components/net/http_loader.rs]: Indeed, its fine here

File: components/constellation/constellation.rs
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: Ahh I was getting confused with freezing pipelines which causes a panic if it gets frozen twice. I don't think there is an issue with closing a pipeline twice.

**Comment by asajeffrey** [comment]
Rebased. @bors-servo try

**Comment by asajeffrey** [comment]
Squashed and rebased @bors-servo r=ConnorGBrewster

**Pull request merged**
